### ISSUE TITLE
Event publish issue between two synced instances

### ISSUE THREAD

**Issue by ghost** [MISP/MISP#5325]
Hi,
I have an issue of my MISP instances about event publishing.

I have two MISP instances A and B. And I set MISP A as a remote server on MISP B for sync.
When I published an event on one server, the event would be synced on the other server.

BUT, it shows the event was not published on the other one.
(No matter I publish the event on A or B, there would be the same issue.)
 * Event on A（ published ）-> Event on B ( Not published )

When I check my log of MISP B ( the server which is shared), the Action was shown as "ADD", not "PUBLISH".

I set these two instances on the same Esxi server, is this the reason?(or any network setting issues?)

I still don't know why it happend, it would be very helpful if anyone can give me some details or comments. **Thank you!**

**Comment by packetwatch** [comment]
I believe that is expected behavior. It should show the creator org on the receiving server as the org on the originating one, but "published" means that particular server offered it up to the distribution group. Your second server did not publish it, only received it. By the fact the originating org is a foreign one, you know it was published by someone else. So "published" isn't reflective of the overall Event status, but rather the status as it pertains to the event on that particular server.

It's against the MISP model, but if you modify the event on the receiving server and change the distribution, you can then publish it from that server as well. 

Hope that helps!

**Comment by ghost** [comment]
Thanks for your comment, packetwatch!

Actually, B is working as a hub and connected with other servers (C,D,E)
At present, the published events could be synced from C/D via the hub B and shown on A as "published". 
For these cases, i checked the log stored in hub B and it shows me the action was "publish".

Also, regarding to the defination of "publish",
I understand it as that if someone publish his event to all, then all connected servers would be able to accept it as published.
In my environment, if I publish an event on A, it should be synced on HUB B as "published". 
And when other servers(such as C,D,E) are connecting to HUB B, they should receive them as "published, too.

This is what I read from MISP RFC. And in this case, if I publish an event on HUB B, it was synced on A as "published" successfully.
I really do not know the reason and it confused me for a while... 

I wonder maybe there would be some setting of sending server?

**Comment by packetwatch** [comment]
I agree, it's confusing. I'm not aware of a setting to retain the published flag, but maybe somebody knows the code better. I'm performing a similar function with Sharing Groups and having similar trouble. I can publish and receive fine in a spoke-and-hub configuration as long as it's the hub that publishes, but not if it comes from one of the remote instances. In my case, having direct sync connections works fine, but that is not an option for these particular instances, so I was relying on sharing groups. Sounds like we have the same general issue. I opened #5322 because I thought it may be something to do with the distribution level. I read somewhere the level drops by one with every sync? 4-3-2-1 or something...

**Comment by ghost** [comment]
Sorry for my late reply.

Actually, I changed my server settings and it kind of works on my problem.
I mean, it might work only if you are setting the distribution level from 0-3(not Sharing Group)

In my case, I changed the setting of "MISP.delegation" to be true.
From the description of it, it seems to allow owner organisation(Who create this event) to remain anonymous and ask another organisation( synced one / shared one) to publish the event for them.

Because after I changed the setting, it works for me, I think it might be working as the following steps:
1. Create an event on A and publish it.
2. (Share the published event to synced server(Like B) and ask them to publish it)
3. Received the published event on B and publish it as B.

ONLY in this case, the event would be shown "published" when it was created and published by a remote instance.

Also as you mentioned, the distribute level will drop by one with every sync. But only if it is 1 or 2.
Like when it is 2, Connected communities -> This communitiy only -> Your organisation only.
When it is 1, This communitiy only -> Your organisation only

I tried to set it as sharing group but it didn't work on Sharing group.
I will try to survey the difference.

**Comment by iglocska** [comment]
If you ever see MISP A -> MISP B pushing data from A to B but data B not publishing the data, it's a smell of the following issues:

1. Background workers are not running / dead on B
2. The sync user on MISP B doesn't have publishing rights.

Could you double check these?

**Comment by ghost** [comment]
@iglocska,
Thanks for your answer.

Yes, I checked both of the issues you mentioned above.

When I create and publish an event, I check background workers of both A and B are runnning.
And I check the log to make sure the event has been published on A (where I created and published).
Sometimes, I push A to B manually to make sure the event should be pushed to B.

Also, the user I used for sync on MISP B is a orgadmin user, and I made sure it has publishing right.

Based on those actions, the event created on A was still unpublished on B.
Until I changed the server settings of MISP.delegation.

But, synced event is still not published when I set the distribution as a sharing group(only for A and B). Event I have the same server setting and did the same steps.

If you have any comments or advises, it would be very appreciated!

**Comment by iglocska** [comment]
The delegation flag should have absolutely no effect on this, so I'm a bit
stumped.

Just to clarify, the org admin user also has the "sync" permission right?

**Comment by ghost** [comment]
@iglocska,

Yes, the org admin user has the sync permission.

In this case, I double check both of A and B's log.
The only difference is that if B (who receive the shared event) publish the event (event published) or just add the event (event unpublished).

When i set the delegation flag up, it worked and I think maybe it makes A give B a request to publish it.
If it doesn't have a function like that, could you have a check please?
Many thanks for your help.

**Comment by iglocska** [comment]
The delegation is really something pretty simplistic that shouldn't affect
this in any way, I don't see how it would. Really strange.

**Comment by ghost** [comment]
@iglocska

I check my MISP version which is 2.4.102 and asked other friends who have no problem on this issue, they are using MISP 2.4.116.

I'm going to update my MISP to the latest version and delegation setting's modification is being a temporary solution for me.

Plan to validate again after update.Thanks for your comments.

**Comment by holidayz1** [comment]
+1

**Comment by holidayz1** [comment]
on version v2.4.118

**Comment by ghost** [comment]
@holidayz1 
I updated my instance A to v2.4.118 without any setting modification except the delegation I mentioned above.
Maybe you can have a try if the delegation setting would work on your instance or not. :>

**Comment by simon-willgoss** [comment]
@YifanWang-data - have you configured the sync for PUSH and/or PULL?

**Comment by ghost** [comment]
@simon-willgoss 
I did set some push/pull filter on A for syncing to B before I found the issue.
But the issue was not that it could not be pushed to B. The issue was that it was pushed but without "published" value.

And as I mentioned, I modified the delegation setting and it worked for me, even though I haven't known the reason.
